many problems fixed and new features

AlexMSatalia opened the pull request:
Upgraded error managing
add remove_constraint to satmodel
remove print_problem and changed get_file_str to build_file_model
corrected variable privacy
add print_constraints
put methods on static when possible
add get_variable functions
put add_var on 'private' (so that users have to use add_continuous_var)
of course, corrected continious ;)
add var.update_name
add build_from_file to build a SATmodel with a given file

AlexMSatalia committed:
many problems fixed and new features

stanle commented on solveengine/basemodel.py:
```diff
@@ -33,31 +34,32 @@ class BaseModel(object):
+        self.__options = BaseModel.OPTIONS(sleeptime, debug)
+        self.__solver_status = str(SolverStatusCode.NOTSTARTED)
+        self.__se_status = str(SEStatusCode.NOTSTARTED)
+        #self.run_time_limit = run_time_limit
```
why is this commented?
unused argument?

AlexMSatalia replied:
preparation for later argument : optional timeout
but you are right to spot it, I have actually now removed it

stanle commented on solveengine/examples/usage.py:
```diff
@@ -8,7 +8,7 @@
 # the token
 token = "2prMEegqCAFB5eqmfuXmGufyJ+PcMzJbZaQcvqrhtx4="
-staticmethod
+
```
seems you did not run this as the method names in this files have not be corrected

AlexMSatalia replied:
that's weird, it should have been removed, I'll double check for the next commit

stanle commented on solveengine/mipmodel.py:
```diff
@@ -8,8 +8,10 @@
 
 from enum import Enum
 from collections import namedtuple
+from numpy import ndarray
```
do we really need this dependency? The other way of modelling does not need this. 
(see below)

at the very least it should be in the requirements.txt

stanle commented on solveengine/mipmodel.py:
```diff
@@ -220,7 +254,7 @@ def set_to_maximize(self):
-                            Aeq = None, beq = [], 
+                            Aeq=None, beq=[],
                             lb=[], ub=[], 
                             int_list=[], bin_list=[]):
```
lists are unsafe as default values because they are mutable, us tuples instead
i.e. this code
```
def t(a=[]):
    a.extend([1])
    print(a)
t()
t()
```
prints
```
[1]
[1, 1]
```

AlexMSatalia replied:
Wow, did not see thats, thats interesting.. Will figure out a solution (tuple or maybe a None default value)

AlexMSatalia replied:
does this look nice enough ?

def t(a=None):
         if a is None: a = list()

stanle replied:
that would work,
you can also do
```
a = a or list()
```

AlexMSatalia replied:
looks so nice !

stanle replied:
the only problem is when somebody uses a custom class as input which somehow
simulates an iterable by implementing `__getitem__` (which is ok) and at the same time implements `__bool__` which returns `False`. 

https://docs.python.org/3/library/stdtypes.html
see 4.1

stanle replied:
but how would do such a thing?

AlexMSatalia replied:
who* would do such a thing.. lets do 
a = list() if a is None else a

stanle commented on solveengine/mipmodel.py:
```diff
@@ -648,3 +644,73 @@ def __imul__(self, other):
+                   name='b', type_=list)
+
+    nb_vars = len(f)
+    if len(A.shape) != 2:
```
I do not see why we restrict this to be ndarrays?
we can achieve what we want using the methods:
A[x][y] 
len(A)
len(A[x])
which is supported by ndarray and list of lists
it can also be easily supported by any custom class which calculates these values on the fly
hence I would not type check this and any of the other inputs

the pythonic programming style determines an object's type by inspection of its method or attribute signature rather than by explicit relationship to some type object. Typecheck, for me, only make sense for custom/local types.

AlexMSatalia replied:
Should have checked about the A[x][y] possibility before, sorry
This should then be feasible with a function that will check the 'matrix' callability

Commit:
fixed spelling bug

Commit:
added missing newline when printing interactive

AlexMSatalia committed:
update examples, satsfied reviewing (check_matrix)

AlexMSatalia committed:
corrected few things

AlexMSatalia committed:
update changes before pushing back
Merge branch 'Bugfix/new_version' of github.com:Satalia/solveengine-python-modeling-lib into Bugfix/new_version

AlexMSatalia committed:
many updates, cant remember because of git problem

AlexMSatalia committed:
Merge branch 'Bugfix/new_version' of github.com:Satalia/solveengine-python-modeling-lib into Bugfix/new_version
